api/src/Rbac/Access/Gate.php

```diff
@@ -0,0 +1,237 @@
+<?php
+
+/*
+ * This file is part of the Spira framework.
+ *
+ * @link https://github.com/spira/spira
+ *
+ * For the full copyright and license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace Spira\Rbac\Access;
+
+use Illuminate\Contracts\Auth\Access\Gate as GateContract;
+use Illuminate\Contracts\Auth\Authenticatable;
+use Spira\Contract\Exception\NotImplementedException;
+use Spira\Rbac\Item\Item;
+use Spira\Rbac\Item\Rule;
+use Spira\Rbac\Storage\StorageInterface;
+
+class Gate implements GateContract
+{
+    const GATE_NAME = 'spira.rbac';
+
+    /**
+     * @var StorageInterface
+     */
+    private $storage;
+
+    /**
+     * @var callable
+     */
+    private $userResolver;
+
+    /**
+     * @var array
+     */
+    private $defaultRoles;
+
+    protected static $itemCache = [];
+
+    protected static $ruleCache = [];
+
+    /**
+     * @param StorageInterface $storage
+     * @param callable $userResolver
+     * @param array $defaultRoles
+     */
+    public function __construct(StorageInterface $storage, callable $userResolver, array $defaultRoles = [])
+    {
+        $this->storage = $storage;
+
+        $this->userResolver = function () use ($userResolver) {
+
+            $user = call_user_func($userResolver);
+            if (! $user) {
+                throw new UserNotFoundException;
+            }
+
+            return $user;
+        };
+
+        $this->defaultRoles = $defaultRoles;
+    }
+
+    /**
+     * Determine if a given ability has been defined.
+     *
+     * @param  string $ability
+     * @return bool
+     */
+    public function has($ability)
+    {
+        throw new NotImplementedException;
+    }
+
+    /**
+     * Define a new ability.
+     *
+     * @param  string $ability
+     * @param  callable|string $callback
+     * @return GateContract
+     */
+    public function define($ability, $callback)
+    {
+        throw new NotImplementedException;
+    }
+
+    /**
+     * Define a policy class for a given class type.
+     *
+     * @param  string $class
+     * @param  string $policy
+     * @return GateContract
+     */
+    public function policy($class, $policy)
+    {
+        throw new NotImplementedException;
+    }
+
+    /**
+     * Determine if the given ability should be granted.
+     *
+     * @param  string $itemName
+     * @param  array|mixed $arguments
+     * @return bool
+     */
+    public function check($itemName, $arguments = [])
+    {
+        if (! ($item = $this->getItem($itemName))) {
+            return true;
+        }
+
+        try {
+            //default roles check
```

Comment:
Please extend this comment to explain the scenario where default roles are used.